mesonbuild/interpreter/interpreter.py

```diff
@@ -130,6 +131,58 @@
 
     BuildTargetSource = T.Union[mesonlib.FileOrString, build.GeneratedTypes, build.StructuredSources]
 
+# This is not comprehensive; it only contains the tools to help detect the
+# combinations that might end up being used, perhaps unknowingly by the user,
+# but which are known to cause problems. MSVC tools can easily be mistakenly
+# mixed with the llvm/clang tools, if not by the user, then by meson itself.
+_TOOLCHAIN_COMPAT_FAMILIES = {
+    'llvm': ({'clang-cl', 'clang'}, {'llvm-lib', 'llvm-ar'}, {'lld-link', 'ld.lld'}),
+    'msvc': ({'cl'}, {'lib'}, {'link'}),
+}
+
+# If the user's enabled 'b_lto' and are using a supporting compiler/linker on
+# windows (i.e. clang/ld.lld, clang-cl/lld-link), then unless they also specify
+# a suitable static linker (llvm-ar or llvm-lib), meson unfortunately tries to
+# quietly go and guess at a static lib/linker tool to use. Under the normal
+# visual studio native tools environment, it picks the 'lib' tool, which results
+# in an unhelpful error -
+#   "invalid or corrupt file: cannot read at 0x..."
+# We can provide a much more helpful error to the user.
+#
+# More generally, however, we can do this kind of tool combination problem
+# checking here.  E.g. we don't expect any of the clang/llvm tools to be mixed
+# with the msvc tools like 'cl', 'lib', 'link'.
+#
+# "Why even bother with this at all?  If the user's mixing weird combinations of
```

Comment:
Random quote

Reply:
Random statement 😄 

To repeat my rambling PR description point on this big comment block in particular -
>  I'll trim down that huge comment if we settle an approach

so that quoted question ("Why even bother ...") is my way of presenting and addressing a possible argument against having this piece of validation that tries to help users from shooting themselves (intentionally or accidentally) in the foot.

"Why not just make a statement without posing a question like this?", you might wonder 😄   It's just a simple, effective way of walking the reader through a thought process involving pros and cons to an approach, in an area where I'm also trying to draw attention to what I think is perhaps an unfortunate choice of meson to employ a kind of "quiet, search and guess"-like behaviour that I've fallen foul of in a few areas now, which I think is probably less useful in the long term than a kind of simpler "stop, report, and require explicit user choice"-like principle.